crates/store/src/state.rs

```diff
@@ -96,17 +96,113 @@ pub struct TransactionInputs {
     pub found_unauthenticated_notes: BTreeSet<NoteId>,
 }
 
+/// Blockchain wrapper for `Mmr` .c
+#[derive(Debug, Clone)]
+pub struct Blockchain(Mmr);
+
+impl Blockchain {
+    /// Returns a new Blockchain.
+    pub fn new(chain_mmr: Mmr) -> Self {
+        Self(chain_mmr)
+    }
+
+    /// Returns the tip of the chain, i.e. the number of the latest block in the chain.
+    pub fn chain_tip(&self) -> BlockNumber {
+        let block_number: u32 = (self.0.forest() - 1)
+            .try_into()
+            .expect("chain_mmr always has, at least, the genesis block");
+
+        block_number.into()
+    }
+
+    /// Returns the chain length.
+    pub fn chain_length(&self) -> BlockNumber {
+        let latest = self.chain_tip();
+        latest + 1
+    }
+
+    /// Returns the current peaks of the MMR.
+    pub fn peaks(&self) -> MmrPeaks {
+        self.0.peaks()
+    }
+
+    /// Returns the peaks of the MMR at the state specified by `forest`.
+    ///
+    /// # Errors
+    /// Returns an error if the specified `forest` value is not valid for this MMR.
+    pub fn peaks_at(&self, forest: usize) -> Result<MmrPeaks, MmrError> {
+        self.0.peaks_at(forest)
+    }
+
+    /// Adds a block commitment to the MMR. The caller must ensure that this commitent is the one
+    /// for the next block in the chain.
+    pub fn add(&mut self, block_commitment: RpoDigest) {
+        self.0.add(block_commitment);
+    }
+
+    /// Returns an [`MmrProof`] for the leaf at the specified position.
+    pub fn open(&self, pos: usize) -> Result<MmrProof, MmrError> {
+        self.0.open_at(pos, self.0.forest())
+    }
+
+    /// Returns a reference to the underlying [`Mmr`].
+    pub fn as_mmr(&self) -> &Mmr {
+        &self.0
+    }
+
+    /// Returns the latest block number and partial mmr.
+    pub fn partial_mmr_from_blocks(
+        &self,
+        blocks: &BTreeSet<BlockNumber>,
+        latest_block_number: BlockNumber,
+    ) -> Result<PartialMmr, GetBatchInputsError> {
+        // Grab the block merkle paths from the inner state.
+        let partial_mmr = {
+            let latest_block_num = latest_block_number;
+
+            // Using latest block as the target forest means we take the state of the MMR one before
+            // the latest block. This is because the latest block will be used as the reference
+            // block of the batch and will be added to the MMR by the batch kernel.
+            let target_forest = latest_block_num.as_usize();
```

Comment:
```suggestion
            // Using latest block as the target forest means we take the state of the MMR one before
            // the latest block. This is because the latest block will be used as the reference
            // block of the batch and will be added to the MMR by the batch kernel.
            let target_forest = latest_block_number.as_usize();
```